modules/stream/src/main/scala/ch/epfl/bluebrain/nexus/sourcing/persistence/Projections.scala

```diff
@@ -0,0 +1,214 @@
+    * @return a future () value
+    */
+  def recordProgress(id: String, progress: ProjectionProgress): F[Unit]
+
+  /**
+    * Retrieves the progress for the specified projection projectionId. If there is no record of progress
+    * the [[ch.epfl.bluebrain.nexus.sourcing.persistence.ProjectionProgress.NoProgress]] is returned.
+    *
+    * @param id an unique projectionId for a projection
+    * @return a future progress value for the specified projection projectionId
+    */
+  def progress(id: String): F[ProjectionProgress]
+
+  /**
+    * Record a specific event against a index failures log projectionId.
+    *
+    * @param id            the project identifier
+    * @param persistenceId the persistenceId to record
+    * @param progress      the progress to record
+    * @param event         the event to be recorded
+    */
+  def recordFailure(id: String, persistenceId: String, progress: ProjectionProgress, event: E): F[Unit]
+
+  /**
+    * An event stream for all failures recorded for a projection.
+    *
+    * @param id the projection identifier
+    * @return a source of the failed events
+    */
+  def failures(id: String): Source[(E, ProjectionProgress), _]
+}
+
+object Projections {
+
+  private class Statements(as: ActorSystem) {
+    private val journalConfig = as.settings.config.getConfig("cassandra-journal")
+    private val cfg           = new CassandraPluginConfig(as, journalConfig)
+
+    val keyspace: String      = cfg.keyspace
+    val progressTable: String = journalConfig.getString("projection-progress-table")
+    val failuresTable: String = journalConfig.getString("projection-failures-table")
+
+    val createKeyspace: String =
+      s"""CREATE KEYSPACE IF NOT EXISTS $keyspace
+         |WITH REPLICATION = { 'class' : ${cfg.replicationStrategy} }""".stripMargin
+
+    val createProgressTable: String =
+      s"""CREATE TABLE IF NOT EXISTS $keyspace.$progressTable (
+         |projectionId varchar primary key, progress text)""".stripMargin
+
+    val createFailuresTable: String =
+      s"""CREATE TABLE IF NOT EXISTS $keyspace.$failuresTable (
+         |projectionId varchar, offset bigint, persistenceId text, progress text, event text,
+         |PRIMARY KEY (projectionId, offset, persistenceId))
+         |WITH CLUSTERING ORDER BY (offset ASC)""".stripMargin
+
+    val recordProgressQuery: String =
+      s"UPDATE $keyspace.$progressTable SET progress = ? WHERE projectionId = ?"
+
+    val progressQuery: String =
+      s"SELECT progress FROM $keyspace.$progressTable WHERE projectionId = ?"
+
+    val recordFailureQuery: String =
+      s"""INSERT INTO $keyspace.$failuresTable (projectionId, offset, persistenceId, progress, event)
+         |VALUES (?, ?, ?, ?, ?) IF NOT EXISTS""".stripMargin
+
+    val failuresQuery: String =
+      s"SELECT progress, event from $keyspace.$failuresTable WHERE projectionId = ? ALLOW FILTERING"
+  }
+
+  private class CassandraProjections[F[_], E](
```

Comment:
Why Encoder and Decoder are in that position and not in `[F[_], E: Encoder: Decoder]` ?

Reply:
No particular reason... left-over from a previous direct use of the instances. :)

Do you think it's worth changing?